core/services/workflows/syncer/engine_registry.go

```diff
@@ -12,64 +13,98 @@ import (
+	metadata GetWorkflowMetadata
+	service  services.Service
+}
+
 type EngineRegistry struct {
-	engines map[string]services.Service
+	engines map[EngineKey]ServiceWithMetadata
 	mu      sync.RWMutex
 }
 
 func NewEngineRegistry() *EngineRegistry {
 	return &EngineRegistry{
-		engines: make(map[string]services.Service),
+		engines: make(map[EngineKey]ServiceWithMetadata),
 	}
 }
 
+// KeyFor generates a key that will be used to identify the engine in the engine registry.
+// This is used instead of a Workflow ID, because the WID will change
+// if the workflow code is modified.
+func (r *EngineRegistry) KeyFor(owner []byte, name string) EngineKey {
+	return EngineKey(hex.EncodeToString(owner) + "-" + name)
+}
+
 // Add adds an engine to the registry.
-func (r *EngineRegistry) Add(id string, engine services.Service) error {
+func (r *EngineRegistry) Add(key EngineKey, engine services.Service, metadata GetWorkflowMetadata) error {
 	r.mu.Lock()
 	defer r.mu.Unlock()
-	if _, found := r.engines[id]; found {
+	if _, found := r.engines[key]; found {
 		return errors.New("attempting to register duplicate engine")
 	}
-	r.engines[id] = engine
+	r.engines[key] = ServiceWithMetadata{
+		metadata: metadata,
+		service:  engine,
+	}
 	return nil
 }
 
 // Get retrieves an engine from the registry.
-func (r *EngineRegistry) Get(id string) (services.Service, error) {
+func (r *EngineRegistry) Get(key EngineKey) (services.Service, GetWorkflowMetadata, error) {
 	r.mu.RLock()
 	defer r.mu.RUnlock()
-	engine, found := r.engines[id]
+	engine, found := r.engines[key]
 	if !found {
-		return nil, errNotFound
+		return nil, GetWorkflowMetadata{}, errNotFound
 	}
-	return engine, nil
+	return engine.service, engine.metadata, nil
+}
+
+// GetAllKeys retrieves all of the keys that are currently known by the engine registry.
+func (r *EngineRegistry) GetAllKeys() []EngineKey {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	keys := []EngineKey{}
+	for key := range r.engines {
+		keys = append(keys, key)
+	}
+	return keys
 }
 
 // Contains is true if the engine exists.
-func (r *EngineRegistry) Contains(id string) bool {
+func (r *EngineRegistry) Contains(id EngineKey) bool {
 	r.mu.RLock()
 	defer r.mu.RUnlock()
 	_, found := r.engines[id]
 	return found
 }
 
 // Pop removes an engine from the registry and returns the engine if found.
-func (r *EngineRegistry) Pop(id string) (services.Service, error) {
+func (r *EngineRegistry) Pop(id EngineKey) (services.Service, GetWorkflowMetadata, error) {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 	engine, ok := r.engines[id]
 	if !ok {
-		return nil, fmt.Errorf("pop failed: %w", errNotFound)
+		return nil, GetWorkflowMetadata{}, fmt.Errorf("pop failed: %w", errNotFound)
 	}
 	delete(r.engines, id)
-	return engine, nil
+	return engine.service, engine.metadata, nil
 }
 
 // PopAll removes and returns all engines.
-func (r *EngineRegistry) PopAll() []services.Service {
+func (r *EngineRegistry) PopAll() ([]services.Service, []GetWorkflowMetadata) {
```

Comment:
why separate the two instead of returning a slice of ServiceWithMetadata?

Reply:
Combined them